ENH:WIP: Add substitution layer

SantiagoTorres opened the pull request:
**Fixes issue #**:

None

**Description of the changes being introduced by the pull request**:

This pull request adds the substitution layer code for our current integrators.
Note that this solution is just a stepping stone into a more robust,
full-fledged solution for a version 0.3.0 of in-toto

    Integrators and package managers that are performing in-toto verification 
need a way to signal some contextual information about the artifacts that 
are being verified (e.g., what's the name of the actual package that was 
received!). The simplest way to provide this functionality is by adding a 
simple replacement layer. Thus, a package manager that wants to verify can call: 
 
   `in_toto_verify(metadata_path, root_layout_keys, param_substitutions=...) `
     
This PR enables this functionality on the root layout, so that artifact 
names or commands to run can be provided to in-toto before execution. 
The parameter dictionary is a simple non-nested {key:value} set of strings 
that can be passed to perform substitution on python-like formatted strings 
within the layout. The tokens to replace are to be defined by the project 
owner when creating the layout. 


**Please verify and check that the pull request fulfills the following
requirements**:

- [x] The code follows the [Code Style Guidelines](https://github.com/secure-systems-lab/code-style-guidelines#code-style-guidelines)
- [x] Tests have been added for the bug fix or new feature
- [ ] Docs have been added for the bug fix or new feature

SantiagoTorres committed:
ENH:verifylib: add support for param substitution

This is a WIP patch, that attempts to add parameter substitution to
in-toto. This is a first attempt at adding a more thorough feature.

SantiagoTorres committed:
FIX: verifylib: the run-field is a list, not a str

SantiagoTorres committed:
FIX:verifylib: add rule substitution on inspection

SantiagoTorres committed:
MINOR:verifylib: paramsub: fix typo in warning msg

SantiagoTorres committed:
ENH:test: add parameter substitution tests.

vladimir-v-diaz commented:
Is this pull request still a work in progress (as indicated in the title)?

Can you also please add an example in the summary of how you expect users to use parameter substitution with in-toto?

vladimir-v-diaz commented on in_toto/verifylib.py:
```diff
@@ -48,6 +49,35 @@
+def _sanitize_parameter_dictionary(parameter_dictionary):
+  """
+  <Purpose>
+    Internal dictionary that ensures that the parameter dictionary:
```
Should this be "Internal function" ?  I'm not sure what is meant by an internal dictionary.

vladimir-v-diaz commented on in_toto/verifylib.py:
```diff
@@ -48,6 +49,35 @@
+  for key in parameter_dictionary:
+    in_toto.formats.PARAMETER_DICTIONARY_KEY.check_match(key)
+    in_toto.formats.PARAMETER_DICTIONARY_VALUE.check_match(parameter_dictionary[key])
+
```
Is this checking that the dict keys and values are a specific format? 
Why not use SCHEMA.DictOf(), which does exactly that?   See the following example:

https://github.com/in-toto/in-toto/blob/a14f79bf8b81728d4a2e2c1a8795f3ec42dc2c96/in_toto/formats.py#L39-L41

SantiagoTorres replied:
Good catch. I refactored this function out, as it appears it just does what `schema.DictOf` already does. Thanks!

vladimir-v-diaz commented on in_toto/verifylib.py:
```diff
@@ -48,6 +49,35 @@
+      The dictionary to verify.
+
+  <Exceptions>
+    SchemaMismatchError if the dictionary doesn't comply with the requirements
```
Is this supposed to be FormatError?

vladimir-v-diaz commented on in_toto/verifylib.py:
```diff
@@ -240,6 +270,64 @@ def verify_layout_expiration(layout):
+            A dictionary containing key-value pairs for substitution.
+
+  <Exceptions>
+    securesystemslib.FormatException:
```
I'm not sure this is the correct exception.  Isn't it FormatError?  Please verify that the expected exceptions are documented correctly.  The _sanitize_parameter_dictionary() call on line 295 is expected to raise SchemaMismatchError (according to its docstring), but that's not documented here.

SantiagoTorres replied:
Yes, I left this as a placeholder. Let me make sure I have the name right.

vladimir-v-diaz commented on in_toto/verifylib.py:
```diff
@@ -240,6 +270,64 @@ def verify_layout_expiration(layout):
+  """
+  _sanitize_parameter_dictionary(parameter_dictionary)
+
+  for step in layout.steps:
```
Why isn't the `layout` argument validated like `parameter_dictionary`?  Is it assumed, and why isn't that explicitly stated in the docstring?

SantiagoTorres replied:
I'm not completely sure, this is a PR that adds `parameter_dictionary`. I'll dig up on that on a separate PR (I'm almost sure that it's checked at the time of de-serialization somewhere else in the code).

vladimir-v-diaz commented on in_toto/verifylib.py:
```diff
@@ -240,6 +270,64 @@ def verify_layout_expiration(layout):
+    ]
+
+    new_run = [x.format(**parameter_dictionary) for x in inspection.run]
+    inspection.run = new_run
```
Per our guidelines against list comprehensions, please consider using s.replace() instead, and being explicit about what is being replaced.

I suggest that the `parameter_dictionary` argument be of the form:
```
{"{COMMAND}", "vim"}
```
It might be easier for others to see that what you want to replace is `{COMMAND}`  and not `COMMAND`.

SantiagoTorres replied:
This is how it works right now. That's why we are using the built-in `format` call, rather than `replace`

vladimir-v-diaz replied:
Let me use the unit test example to be clear.
```
def test_substitute(self):
"""Do a simple substitution on the expected_command field"""
substitute_parameters(self.layout, {"EDITOR": "vim"})
```

I think it might be more readable if the second argument to `substitute_parameters()` was required as follows:
```Python
{"{EDITOR}", "vim"}
```

Just a suggestion.

SantiagoTorres replied:
Hmm, I'm not completely convinced to having the argument named like that. It just adds complexity on the codebase and I think the change in clarity is very little. We can discuss offline to decide something.

vladimir-v-diaz commented on in_toto/verifylib.py:
```diff
@@ -240,6 +270,64 @@ def verify_layout_expiration(layout):
+  <Purpose>
+    This function is a transitionary measure for parameter substitution (or
+    any other solution defined by the in-toto team). As of now, it acts as 
+    a very simple replacement layer for python-like parameters
```
It might be good to discuss the use cases for this functionality.  Why and when should an adopter use the substitution layer?  

Also, is this functionality usable from the command-line tools?  This pull request appears to only support it from `verifylib.py`.

SantiagoTorres replied:
No, I didn't add the functionality from the command line tools yet, as the only people interested in this (potentially temporary) feature is doing a direct call to verifylib.

vladimir-v-diaz commented on tests/test_param_substitution.py:
```diff
@@ -0,0 +1,201 @@
+  See LICENSE for licensing information.
+
+<Purpose>
+  Test verifylib functions.
```
This purpose seems wrong.  It's not really testing the verifylib functions, but only param substitution.  There already is a test_verifylib.py.  Shouldn't the substitute function be tested there, in `test_verifylib.py`?

SantiagoTorres replied:
I did this on purpose, as to avoid a feature that we may later migrate of. I.e., it makes easier to refactor the code this way

vladimir-v-diaz reviewed (commented):
Okay, here are my remaining comments for this pull request.

SantiagoTorres commented:
I see most of the feedback is about documenting the feature itself. Let me take care of this and post it here. In the meantime, I'll fix one of the issues and remove the WIP label

SantiagoTorres committed:
FIX:verifylib: refactor parameter sub schema obj

The schema object as provided by formats contains facilities to define a
dictionary of arbitrary key schemas and arbitrary value schemas. Use
that object instead of the home-brewed check function to reduce
verbosity and maintainance overhead.

SantiagoTorres committed:
FIX:verifylib:param_subs:inline list comprehension

SantiagoTorres commented:
Note: builds will fail until we fix https://github.com/secure-systems-lab/securesystemslib/issues/153

vladimir-v-diaz commented:
The master/develop branch seems to be passing.  Has this pull request been sync'd with the develop branch?

SantiagoTorres committed:
FIX:param_sub: address review comments.

This commit mainly fixes wrong docstrings and rebases with master, which
changed the method name for `import_rsa_keys...` to
`import_public_keys...`

vladimir-v-diaz commented on in_toto/verifylib.py:
```diff
@@ -240,6 +242,84 @@ def verify_layout_expiration(layout):
+    The layout object will have any tags replaced with the corresponding
+    values defined in the parameter dictionary.
+  """
+  in_toto.formats.PARAMETER_DICTIONARY_SCHEMA.check_match(parameter_dictionary)
```
nit: Consider including an example of a parameter_dictionary in a comment (especially since the strings being replaces are expected to contain the curly braces).  I had to refer back to the unit tests and in_toto.formats to confirm the expected format.

vladimir-v-diaz commented:
I tried executing some of the functions and unit tests and came across the following error (it might be related to GPG?):

```Bash
======================================================================
ERROR: test_verify_link_signature_thresholds__S_M_M (tests.test_verifylib.TestInTotoVerifyThresholdsGpgSubkeys)
Allowed trust delegation.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/vlad/projects/in-toto/tests/test_verifylib.py", line 1423, in test_verify_link_signature_thresholds__S_M_M
    verify_link_signature_thresholds(layout, chain_link_dict)
  File "/Users/vlad/projects/in-toto/in_toto/verifylib.py", line 481, in verify_link_signature_thresholds
    " found '{}'".format(step.threshold, valid_authorized_links_cnt))
in_toto.exceptions.ThresholdVerificationError: Step requires at least '1' links validly signed by different authorized functionaries. Only found '0'
```

vladimir-v-diaz commented on tests/test_param_substitution.py:
```diff
@@ -0,0 +1,215 @@
+
+  def test_substitute_no_var(self):
+    """Raise an error if the parameter is not filled-in"""
+    with self.assertRaises(KeyError):
```
Is there a unit test for a parameter dictionary that doesn't contain the expected format?

Another thing you  might want to do is raise a warning if the rule to be substituted doesn't contain the curly braces, which someone might forget to include.

vladimir-v-diaz commented on in_toto/verifylib.py:
```diff
@@ -240,6 +242,84 @@ def verify_layout_expiration(layout):
+    new_material_rules = []
+    for rule in step.expected_materials:
+      new_rule = []
+      for stanza in rule:
```
Why did you choose the name `stanza`?  Is this terminology used when dealing with variable substitutions?

SantiagoTorres replied:
No, but I couldn't think of anything better :confounded:

vladimir-v-diaz commented:
I just had a few more comments for the follow-up review of this pull request.  Feel free to merge when you wish.
